sdk/src/Services/S3/Custom/AmazonS3Client.Extensions.cs

```diff
@@ -564,6 +571,292 @@ public async Task<string> GetPreSignedURLAsync(GetPreSignedUrlRequest request)
+
+            var credentials = Config.DefaultAWSCredentials ?? DefaultIdentityResolverConfiguration.ResolveDefaultIdentity<AWSCredentials>();
+            if (credentials == null)
+                throw new AmazonS3Exception("Credentials must be specified, cannot call method anonymously");
+
+            // Resolve credentials asynchronously
+            var immutableCredentials = await credentials.GetCredentialsAsync().ConfigureAwait(false);
+
+            var irequest = CreateAndProcessRequest(request);
+            return BuildPresignedPostResponse(request, irequest, credentials);
+        }
+
+        /// <summary>
+        /// Marshalls the parameters for a presigned POST request to create a proper IRequest object.
+        /// </summary>
+        /// <param name="createPresignedPostRequest">The presigned POST request</param>
+        /// <returns>Internal request object</returns>
+        private static IRequest MarshallCreatePresignedPost(CreatePresignedPostRequest createPresignedPostRequest)
+        {
+            IRequest request = new DefaultRequest(createPresignedPostRequest, "AmazonS3");
+            request.HttpMethod = "POST";
+
+            // Post uses root resource path
+            request.ResourcePath = "/";
+            request.UseQueryString = false; // POST uses form data, not query string
+
+            return request;
+        }
+
+        /// <summary>
+        /// Builds the policy document JSON string from the request using Utf8JsonWriter.
+        /// This approach follows AWS SDK patterns and is Native AOT compatible.
+        /// </summary>
+        /// <param name="request">The CreatePresignedPostRequest containing the policy conditions.</param>
+        /// <returns>A JSON string representing the policy document.</returns>
+        private string BuildPolicyDocument(CreatePresignedPostRequest request)
+        {
+#if !NETFRAMEWORK
+            using var arrayPoolBufferWriter = new ArrayPoolBufferWriter<byte>();
+            using var writer = new Utf8JsonWriter(arrayPoolBufferWriter);
+#else
+            using var memoryStream = new MemoryStream();
+            using var writer = new Utf8JsonWriter(memoryStream);
+#endif
+            
+            writer.WriteStartObject();
+            writer.WriteString("expiration", request.Expires.Value.ToString("yyyy-MM-ddTHH:mm:ss.fffZ"));
+            writer.WriteStartArray("conditions");
+            
+            // Add bucket condition (required)
+            writer.WriteStartObject();
+            writer.WriteString("bucket", request.BucketName);
+            writer.WriteEndObject();
+            
+            // Add key condition if specified
+            if (!string.IsNullOrEmpty(request.Key))
+            {
+                // Check if key ends with ${filename} and add special handling
+                if (request.Key.EndsWith("${filename}", StringComparison.Ordinal))
+                {
+                    // Extract the prefix before ${filename}
```

Comment:
this is based off of javascripts https://github.com/aws/aws-sdk-js-v3/blob/main/packages/s3-presigned-post/src/createPresignedPost.ts#L82 logic. But as ive been testing i think there might be a bug where this logic should actually be applied to all form fields, and not just the key. i havent made that change here though since im not 100% confident on it https://github.com/aws/aws-sdk-js-v3/issues/7191.

Reply:
even if we do fix it and replace ${filename} in other form fields - the open question is should we do similar logic to how we replace they request.key value?

Reply:
I'm okay waiting to see if we get a customer ask for adding special handling for `${filename}` till we get a customer ask for the feature.